File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        ListNode *root = nullptr;
        //first step make that root not null
        if (list1 && list2) {
            if (list2->val < list1->val) {
                root = list2;
                list2 = list2->next;
            } else {
                root = list1;
                list1 = list1->next;
            }
        } else {
            if (list1) {
                root = list1;
                list1 = list1->next;
            }
            if (list2) {
                root = list2;
                list2 = list2->next;
            }
        }
        if (!root) {
            return nullptr;
        }
        ListNode *curr = root;
        while (list1 && list2) {
            if (list1->val < list2->val) {
                curr->next = list1;
                list1 = list1->next;
            } else {
                curr->next = list2;
                list2 = list2->next;
            }
            curr = curr->next;
        }
        while(list1) {
            curr->next = list1;
            list1 = list1->next;
            curr = curr->next;

        }
        while(list2) {
            curr->next = list2;
            list2 = list2->next;
            curr = curr->next;

        }
        return root;
    }
};

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        #general strategy will be to count the number of nodes with a single pass
        #then subtract n from the total number of nodes
        # that is the number of times we will need to leap forward, while finding our target
        # to remove we need to set the prev to the one being deleteds next. 
        # and cleaning the pointers
        
        if not head:
            return None
        c = head
        num_nodes = 0
        while c:
            c = c.next
            num_nodes+=1

        target_node = num_nodes - n # the 0 indexed from beginning node to murder
        if target_node == 0:
            return head.next

        c = head
        p = None
        node_index = 0
        while c:
            if node_index == target_node:
                p.next = c.next
                c.next = None
                return head
            p = c
            c = c.next
            node_index+=1